File: py/D6.py


def parseInput():
    file = open("../inputs/D6.input", "r")
    groups = []
    contents = file.read()
    for section in contents.split("\n\n"):
        group = []
        for person in section.splitlines():
            group.append(person)
        groups.append(group)
    return groups

def questionsAnswered(groups):
    count = 0
    for group in groups:
        seen = set()
        for person in group:
            for question in person:
                seen.add(question)
        count += len(seen)
    return count

def p1():
    return questionsAnswered(parseInput())

def unamiouslyAnswered(groups):
    count = 0
    for group in groups:
        seen = {}
        for person in group:
            for question in person:
                if question not in seen:
                    seen[question] = 0
                seen[question] += 1
        count += len(dict(filter(lambda kvp: kvp[1] == len(group), seen.items())))
    return count

def p2():
    return unamiouslyAnswered(parseInput())


print(p1())
print(p2())

File: rb/D6.rb
require 'set'

def parseInput()
    groups = []
    current = []
    File.open("../inputs/D6.input", "r") do |f|
      for line in f.each_line
        if line.chomp == ''
          groups.append(current)
          current = []
          next
        end
        current.append(line.chomp)
      end
    end
    groups.append(current)
    return groups
end

def questionsAnswered(group)
  answered = Set[]
  for person in group
    for question in person.each_char
      answered.add(question)
    end
  end
  return answered.count
end

def p1()
  return parseInput().map {|g| questionsAnswered(g) }.sum()
end

def unamiouslyAnswered(group)
  answered = {}
  for person in group
    for question in person.each_char
      answered[question] = 0 if !answered.include? question
      answered[question] += 1
    end
  end
  return answered.select{|key, value| value == group.count}.count
end

def p2()
  return parseInput().map {|g| unamiouslyAnswered(g) }.sum()
end

puts p1()
puts p2()